core/trino-main/src/main/java/io/trino/sql/SensitiveStatementRedactor.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql;
+
+import com.google.inject.Inject;
+import io.trino.FeaturesConfig;
+import io.trino.Session;
+import io.trino.connector.CatalogFactory;
+import io.trino.execution.CreateCatalogTask;
+import io.trino.execution.QueryPreparer.PreparedQuery;
+import io.trino.metadata.CatalogManager;
+import io.trino.security.AccessControl;
+import io.trino.spi.catalog.CatalogName;
+import io.trino.spi.catalog.CatalogProperties;
+import io.trino.spi.connector.ConnectorName;
+import io.trino.sql.tree.AstVisitor;
+import io.trino.sql.tree.CreateCatalog;
+import io.trino.sql.tree.ExecuteImmediate;
+import io.trino.sql.tree.Explain;
+import io.trino.sql.tree.ExplainAnalyze;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.Identifier;
+import io.trino.sql.tree.Node;
+import io.trino.sql.tree.Prepare;
+import io.trino.sql.tree.Property;
+import io.trino.sql.tree.Statement;
+import io.trino.sql.tree.StringLiteral;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static org.weakref.jmx.$internal.guava.collect.ImmutableSet.toImmutableSet;
+
+public class SensitiveStatementRedactor
```

Comment:
I have concerns about this approach. This couples the implementation with the AST structure and visitor infrastructure, and it requires performing actions that should be the responsibility of the analyzer (e.g., the logic in visitCreateCatalog). In particular, the dependency on `CreateCatalogTask.evaluateProperties` is not appropriate. 

Also, this is not sufficient -- what about values in query plans that may be derived from security sensitive properties? E.g., a TableHandle in a plan may need to include the URL/user/password for a database. Running an EXPLAIN would expose those details.

Reply:
> Also, this is not sufficient -- what about values in query plans that may be derived from security sensitive properties? E.g., a TableHandle in a plan may need to include the URL/user/password for a database. Running an EXPLAIN would expose those details.

I tried to reproduce this but I couldn't find a case where `TableHandle` includes catalog properties. Do you have any specific example when this might happen? Also, is my understanding correct that if security-sensitive properties might leak in EXPLAIN output, it is something that unrelated strictly to dynamic catalogs and might happen for static catalogs as well?